runway/hooks/staticsite/auth_at_edge/templates/check_auth/__init__.py

```diff
@@ -0,0 +1,252 @@
+            config.get('token_issuer'),
+            config.get('client_id'))
+
+        return request
+    # We need new authorization. Get the user over to Cognito
+    except Exception: # noqa pylint: disable=broad-except
+        pkce, pkce_hash = generate_pkce_verifier()
+        payload = {
+            'redirect_uri': 'https://%s%s' % (domain_name, config['redirect_path_sign_in']),
+            'response_type': 'code',
+            'client_id': config["client_id"],
+            'state': base64.urlsafe_b64encode(
+                bytes(json.dumps({
+                    "nonce": nonce,
+                    "requestedUri": requested_uri
+                }).encode())
+            ),
+            'scope': " ".join(config['oauth_scopes']),
+            'code_challenge_method': "S256",
+            'code_challenge': pkce_hash
+        }
+        login_query_string = urlencode(payload, quote_via=quote_plus)
+        headers = config.get('cloud_front_headers')
+        headers['location'] = [
+            {
+                'key': 'location',
+                'value': 'https://%s/oauth2/authorize?%s' % (
+                    config['cognito_auth_domain'],
+                    login_query_string
+                )
+            }
+        ]
+        headers['set-cookie'] = [
+            # Set Nonce Cookie
+            {
+                'key': 'set-cookie',
+                'value': 'spa-auth-edge-nonce=%s; %s' % (
+                    quote_plus(nonce),
+                    config.get('cookie_settings', {}).get('nonce')
+                )
+            },
+            # Set PKCE Cookie
+            {
+                'key': 'set-cookie',
+                'value': 'spa-auth-edge-pkce=%s; %s' % (
+                    quote_plus(pkce),
+                    config.get('cookie_settings', {}).get('nonce')
+                )
+            },
+        ]
+
+        # Redirect user to the Cognito Login
+        return {
+            'status': '307',
+            'statusDescription': 'Temporary Redirect',
+            'headers': headers
+        }
+
+
+def generate_pkce_verifier():
+    """Generate the PKCE verification code."""
+    pkce = random_key(PKCE_LENGTH)
+    pkce_hash = hashlib.sha256()
+    pkce_hash.update(pkce.encode('UTF-8'))
+    pkce_hash = pkce_hash.digest()
+    pkce_hash_b64 = base64.b64encode(pkce_hash)
+    decoded_pkce = pkce_hash_b64.decode("UTF-8")
+    decoded_pkce = re.sub(r'\=', '', decoded_pkce)
+    decoded_pkce = re.sub(r'\+', '-', decoded_pkce)
+    decoded_pkce = re.sub(r'\/', '_', decoded_pkce)
+
+    return [pkce, decoded_pkce]
+
+
+def generate_nonce():
+    """Generate a random Nonce token."""
+    return random_key(NONCE_LENGTH)
+
+
+def random_key(length=15):
+    """Generate a random key of specified length from the allowed secret characters.
+
+    Keyword Args:
+        length (int): The length of the random key
+    """
+    return ''.join(random.choice(SECRET_ALLOWED_CHARS) for _ in range(length))
```

Comment:
Let's swap `random.choice` for `secrets.choice` here https://docs.python.org/3/library/secrets.html#random-numbers